File: wrappers/go/zxc_seekable_test.go
/*
ZXC - High-performance lossless compression

Copyright (c) 2025-2026 Bertrand Lebonnois and contributors.
SPDX-License-Identifier: BSD-3-Clause
*/

package zxc

import (
	"bytes"
	"errors"
	"io"
	"os"
	"path/filepath"
	"sync/atomic"
	"testing"
)

func buildSeekableArchive(t *testing.T, payload []byte) string {
	t.Helper()
	dir := t.TempDir()
	in := filepath.Join(dir, "input.bin")
	out := filepath.Join(dir, "archive.zxc")

	if err := os.WriteFile(in, payload, 0o644); err != nil {
		t.Fatalf("write input: %v", err)
	}
	if _, err := CompressFile(in, out, WithSeekable(true), WithChecksum(true)); err != nil {
		t.Fatalf("CompressFile(seekable): %v", err)
	}
	return out
}

// Asserting both verdicts proves the switch reaches C: on, ErrBadChecksum; off,
// the corrupted bytes come back with no error.
func TestSeekableSetChecksum(t *testing.T) {
	// Incompressible on purpose: a RAW block memcpys a flipped byte straight
	// through, so the decode succeeds and only the checksum objects. A
	// compressible payload would pin this test to today's encoder output.
	payload := make([]byte, 256*1024)
	rng := uint32(0x2E5B9A17)
	for i := range payload {
		rng = rng*1103515245 + 12345
		payload[i] = byte(rng >> 16)
	}
	path := buildSeekableArchive(t, payload)

	s, err := Open(path)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	out := make([]byte, 512)
	for _, on := range []bool{true, false, true} {
		if err := s.SetChecksum(on); err != nil {
			t.Fatalf("SetChecksum(%v): %v", on, err)
		}
		if _, err := s.DecompressRange(out, 0, len(out)); err != nil {
			t.Fatalf("DecompressRange with checksum=%v: %v", on, err)
		}
	}
	s.Close()

	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read archive: %v", err)
	}
	if raw[16] != 0 {
		t.Fatalf("block 0 type = %d, want RAW; the LCG payload is no longer incompressible", raw[16])
	}
	// 16-byte file header, then block 0's header (8 bytes), then payload.
	raw[16+8+4] ^= 0xFF
	bad := filepath.Join(t.TempDir(), "corrupt.zxc")
	if err := os.WriteFile(bad, raw, 0o644); err != nil {
		t.Fatalf("write corrupt: %v", err)
	}

	verifying, err := Open(bad)
	if err != nil {
		t.Fatalf("Open(corrupt): %v", err)
	}
	if err := verifying.SetChecksum(true); err != nil {
		t.Fatalf("SetChecksum(true): %v", err)
	}
	_, err = verifying.DecompressRange(out, 0, len(out))
	verifying.Close()
	if !errors.Is(err, ErrBadChecksum) {
		t.Fatalf("verifying: want ErrBadChecksum, got %v", err)
	}

	// Default is off: bytes come back, wrong, with no error.
	skipping, err := Open(bad)
	if err != nil {
		t.Fatalf("Open(corrupt): %v", err)
	}
	n, err := skipping.DecompressRange(out, 0, len(out))
	skipping.Close()
	if err != nil || n != len(out) {
		t.Fatalf("not verifying: want %d bytes, got %d (%v)", len(out), n, err)
	}
	if bytes.Equal(out, payload[:len(out)]) {
		t.Fatalf("not verifying: expected wrong bytes")
	}
}

func TestSeekableOpenAndQuery(t *testing.T) {
	payload := bytes.Repeat([]byte("ZXCseekable_"), 8192) // ~96 KiB
	path := buildSeekableArchive(t, payload)

	s, err := Open(path)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer s.Close()

	if got := s.DecompressedSize(); got != uint64(len(payload)) {
		t.Fatalf("DecompressedSize = %d, want %d", got, len(payload))
	}
	if s.NumBlocks() == 0 {
		t.Fatalf("NumBlocks = 0, want >= 1")
	}

	if cs, ok, err := s.BlockCompressedSize(0); !ok || err != nil || cs == 0 {
		t.Fatalf("BlockCompressedSize(0) = %d ok=%v err=%v", cs, ok, err)
	}
	if ds, ok := s.BlockDecompressedSize(0); !ok || ds == 0 {
		t.Fatalf("BlockDecompressedSize(0) = %d ok=%v", ds, ok)
	}
	if _, ok, err := s.BlockCompressedSize(s.NumBlocks()); ok || err != nil {
		t.Fatalf("BlockCompressedSize(out-of-range) = ok=%v err=%v, want false, nil", ok, err)
	}
}

func TestSeekableForgedGroupIsAnError(t *testing.T) {
	payload := bytes.Repeat([]byte("ZXCseekable_"), 8192)
	arc, err := os.ReadFile(buildSeekableArchive(t, payload))
	if err != nil {
		t.Fatalf("read archive: %v", err)
	}
	probe, err := OpenBytes(arc)
	if err != nil {
		t.Fatalf("OpenBytes: %v", err)
	}
	n := int(probe.NumBlocks())
	probe.Close()
	// Group 0's anchor opens the table, before the 16-byte footer (size, digest).
	table := (n+63)/64*8 + n*4
	anchor := len(arc) - 16 - table
	if arc[anchor] != 16 {
		t.Fatalf("byte %d = %d, not group 0's anchor", anchor, arc[anchor])
	}
	arc[anchor] ^= 0xFF

	s, err := OpenBytes(arc)
	if err != nil {
		t.Fatalf("OpenBytes must not scan the table: %v", err)
	}
	defer s.Close()
	if _, ok, err := s.BlockCompressedSize(0); !ok || !errors.Is(err, ErrInvalidData) {
		t.Fatalf("BlockCompressedSize(0) on a forged group: ok=%v err=%v", ok, err)
	}
}

func TestSeekableDecompressRange(t *testing.T) {
	payload := make([]byte, 64*1024)
	for i := range payload {
		payload[i] = byte(i * 31)
	}
	path := buildSeekableArchive(t, payload)

	s, err := Open(path)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer s.Close()

	// Full range.
	full := make([]byte, len(payload))
	n, err := s.DecompressRange(full, 0, len(payload))
	if err != nil {
		t.Fatalf("DecompressRange full: %v", err)
	}
	if n != len(payload) || !bytes.Equal(full, payload) {
		t.Fatalf("full range mismatch (n=%d)", n)
	}

	// Mid-range.
	const off, length = 1024, 8192
	mid := make([]byte, length)
	n, err = s.DecompressRange(mid, off, length)
	if err != nil {
		t.Fatalf("DecompressRange mid: %v", err)
	}
	if n != length || !bytes.Equal(mid, payload[off:off+length]) {
		t.Fatalf("mid range mismatch (n=%d)", n)
	}
}

func TestSeekableOpenBytes(t *testing.T) {
	payload := bytes.Repeat([]byte("memseekable_"), 4096)
	path := buildSeekableArchive(t, payload)

	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read archive: %v", err)
	}

	s, err := OpenBytes(data)
	if err != nil {
		t.Fatalf("OpenBytes: %v", err)
	}
	defer s.Close()

	if got := s.DecompressedSize(); got != uint64(len(payload)) {
		t.Fatalf("DecompressedSize = %d, want %d", got, len(payload))
	}

	out := make([]byte, len(payload))
	if _, err := s.DecompressRange(out, 0, len(payload)); err != nil {
		t.Fatalf("DecompressRange: %v", err)
	}
	if !bytes.Equal(out, payload) {
		t.Fatalf("payload mismatch after OpenBytes round-trip")
	}
}

func TestSeekableInvalidBytes(t *testing.T) {
	if _, err := OpenBytes(make([]byte, 32)); err == nil {
		t.Fatalf("OpenBytes on garbage should fail")
	}
	if _, err := OpenBytes(nil); err == nil {
		t.Fatalf("OpenBytes(nil) should fail")
	}
}

// countingReaderAt wraps bytes.Reader and counts ReadAt invocations so
// tests can assert lazy I/O at open and per-block reads.
type countingReaderAt struct {
	inner io.ReaderAt
	calls int64
}

func (c *countingReaderAt) ReadAt(p []byte, off int64) (int, error) {
	atomic.AddInt64(&c.calls, 1)
	return c.inner.ReadAt(p, off)
}

func TestSeekableOpenReader(t *testing.T) {
	payload := make([]byte, 256*1024)
	for i := range payload {
		payload[i] = byte(i*7) ^ 0x5A
	}
	path := buildSeekableArchive(t, payload)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read archive: %v", err)
	}

	cr := &countingReaderAt{inner: bytes.NewReader(data)}
	s, err := OpenReader(cr, int64(len(data)))
	if err != nil {
		t.Fatalf("OpenReader: %v", err)
	}
	defer s.Close()

	// open_reader should have done exactly 3 reads: header, footer, EOF/SEK
	// block headers.
	if got := atomic.LoadInt64(&cr.calls); got != 3 {
		t.Fatalf("open phase calls = %d, want 3", got)
	}

	if got := s.DecompressedSize(); got != uint64(len(payload)) {
		t.Fatalf("DecompressedSize = %d, want %d", got, len(payload))
	}

	// Full-range round-trip.
	out := make([]byte, len(payload))
	if _, err := s.DecompressRange(out, 0, len(payload)); err != nil {
		t.Fatalf("DecompressRange full: %v", err)
	}
	if !bytes.Equal(out, payload) {
		t.Fatalf("payload mismatch after full DecompressRange")
	}

	// Sub-range within a single block: its seek table entries, then the block.
	before := atomic.LoadInt64(&cr.calls)
	chunk := make([]byte, 1024)
	if _, err := s.DecompressRange(chunk, 100, 1024); err != nil {
		t.Fatalf("DecompressRange sub: %v", err)
	}
	if !bytes.Equal(chunk, payload[100:1124]) {
		t.Fatalf("sub-range mismatch")
	}
	if delta := atomic.LoadInt64(&cr.calls) - before; delta != 2 {
		t.Fatalf("single-block sub-range should trigger 2 reads (entries, block), got %d", delta)
	}
}

// hookReaderAt runs hook once, on its next read.
type hookReaderAt struct {
	inner io.ReaderAt
	hook  func()
}

func (h *hookReaderAt) ReadAt(p []byte, off int64) (int, error) {
	if f := h.hook; f != nil {
		h.hook = nil
		f()
	}
	return h.inner.ReadAt(p, off)
}

// From ReadAt all is refused but a range under the getter, which shares nothing.
func TestSeekableReentryFromReader(t *testing.T) {
	payload := bytes.Repeat([]byte("ZXCseekable_"), 8192)
	arc, err := os.ReadFile(buildSeekableArchive(t, payload))
	if err != nil {
		t.Fatalf("read archive: %v", err)
	}
	rd := &hookReaderAt{inner: bytes.NewReader(arc)}
	s, err := OpenReader(rd, int64(len(arc)))
	if err != nil {
		t.Fatalf("OpenReader: %v", err)
	}
	defer s.Close()

	dst := make([]byte, len(payload))
	calls := map[string]func() error{
		"BlockCompressedSize": func() error { _, _, err := s.BlockCompressedSize(0); return err },
		"DecompressRange":     func() error { _, err := s.DecompressRange(dst, 0, len(dst)); return err },
	}
	actions := map[string]func() error{
		"Close":           s.Close,
		"SetDict":         func() error { return s.SetDict([]byte("x"), nil) },
		"DecompressRange": func() error { _, err := s.DecompressRange(make([]byte, 16), 0, 16); return err },
	}
	cases := []struct {
		call, action string
		want         error
	}{
		{"BlockCompressedSize", "Close", ErrSeekableInUse},
		{"BlockCompressedSize", "SetDict", ErrSeekableInUse},
		{"BlockCompressedSize", "DecompressRange", nil},
		{"DecompressRange", "Close", ErrSeekableInUse},
		{"DecompressRange", "SetDict", ErrSeekableInUse},
		{"DecompressRange", "DecompressRange", ErrSeekableInUse},
	}
	for _, c := range cases {
		got := errors.New("hook not run")
		rd.hook = func() { got = actions[c.action]() }
		if err := calls[c.call](); err != nil {
			t.Fatalf("%s with %s from ReadAt: %v", c.call, c.action, err)
		}
		if got != c.want {
			t.Fatalf("%s from ReadAt during %s: %v, want %v", c.action, c.call, got, c.want)
		}
	}
	if n, err := s.DecompressRange(dst, 0, len(dst)); err != nil || !bytes.Equal(dst[:n], payload) {
		t.Fatalf("handle after the refusals: n=%d err=%v", n, err)
	}
}

func TestSeekableOpenReaderRejectsNilAndZero(t *testing.T) {
	if _, err := OpenReader(nil, 100); err == nil {
		t.Fatalf("OpenReader(nil, 100) should fail")
	}
	if _, err := OpenReader(bytes.NewReader([]byte{1, 2, 3}), 0); err == nil {
		t.Fatalf("OpenReader(r, 0) should fail")
	}
}

func TestSeekableOpenReaderRejectsGarbage(t *testing.T) {
	// 64 bytes of zeros pass the minimum-size check but parse as invalid.
	// The cgo handle must still be released (no leak in steady state).
	garbage := bytes.NewReader(make([]byte, 64))
	if _, err := OpenReader(garbage, 64); err == nil {
		t.Fatalf("OpenReader on garbage should fail")
	}
}

func TestSeekTableSizeAndWrite(t *testing.T) {
	compSizes := []uint32{128, 256, 200, 4}
	sz := SeekTableSize(uint64(len(compSizes)))
	if sz == 0 {
		t.Fatalf("SeekTableSize = 0")
	}
	buf := make([]byte, sz)
	n, err := WriteSeekTable(buf, compSizes)
	if err != nil {
		t.Fatalf("WriteSeekTable: %v", err)
	}
	if n != sz {
		t.Fatalf("WriteSeekTable wrote %d bytes, want %d", n, sz)
	}
}

File: wrappers/nodejs/test/zxc.seekable.test.js
/*
 * ZXC - High-performance lossless compression
 *
 * Copyright (c) 2025-2026 Bertrand Lebonnois and contributors.
 * SPDX-License-Identifier: BSD-3-Clause
 */

"use strict";

const zxc = require("../lib/index");

// =============================================================================
// Seekable random-access decompression
// =============================================================================

function buildPayload(size) {
  const buf = Buffer.alloc(size);
  for (let i = 0; i < size; i++) buf[i] = (i * 31) & 0xff;
  return buf;
}

function buildSeekableArchive(payload) {
  return zxc.compress(payload, { seekable: true, checksum: true });
}

describe("Seekable: queries", () => {
  const payload = buildPayload(64 * 1024);
  const compressed = buildSeekableArchive(payload);

  test("reports total decompressed size and at least one block", () => {
    const s = new zxc.Seekable(compressed);
    try {
      expect(s.decompressedSize()).toBe(payload.length);
      expect(s.numBlocks()).toBeGreaterThanOrEqual(1);
    } finally {
      s.close();
    }
  });

  test("reports per-block sizes and rejects out-of-range indices", () => {
    const s = new zxc.Seekable(compressed);
    try {
      const numBlocks = s.numBlocks();
      expect(s.blockCompressedSize(0)).toBeGreaterThan(0);
      expect(s.blockDecompressedSize(0)).toBeGreaterThan(0);
      expect(s.blockCompressedSize(numBlocks)).toBeNull();
      expect(s.blockDecompressedSize(numBlocks)).toBeNull();
    } finally {
      s.close();
    }
  });

  test("block indices are 64-bit, but must stay exact JS integers", () => {
    const s = new zxc.Seekable(compressed);
    try {
      // Above 2^32: no longer truncated into range, just out of range.
      expect(s.blockCompressedSize(2 ** 40)).toBeNull();
      expect(s.blockDecompressedSize(Number.MAX_SAFE_INTEGER)).toBeNull();
      for (const bad of [-1, 1.5, 2 ** 53, Infinity, NaN]) {
        expect(() => s.blockCompressedSize(bad)).toThrow(RangeError);
        expect(() => s.blockDecompressedSize(bad)).toThrow(RangeError);
      }
    } finally {
      s.close();
    }
  });
});

describe("Seekable: decompressRange", () => {
  const payload = buildPayload(64 * 1024);
  const compressed = buildSeekableArchive(payload);

  test("full-range round trip", () => {
    const s = new zxc.Seekable(compressed);
    try {
      const out = s.decompressRange(0, payload.length);
      expect(out.length).toBe(payload.length);
      expect(out.equals(payload)).toBe(true);
    } finally {
      s.close();
    }
  });

  test("mid-range slice is byte-exact", () => {
    const s = new zxc.Seekable(compressed);
    try {
      const off = 1024;
      const len = 8192;
      const out = s.decompressRange(off, len);
      expect(out.length).toBe(len);
      expect(out.equals(payload.subarray(off, off + len))).toBe(true);
    } finally {
      s.close();
    }
  });

  test("zero-length range returns empty buffer", () => {
    const s = new zxc.Seekable(compressed);
    try {
      const out = s.decompressRange(0, 0);
      expect(out.length).toBe(0);
    } finally {
      s.close();
    }
  });
});

describe("Seekable: error handling", () => {
  test("constructor throws on garbage buffer", () => {
    expect(
      () => new zxc.Seekable(Buffer.from([1, 2, 3, 4, 5, 6, 7, 8])),
    ).toThrow();
  });

  test("constructor throws on empty buffer", () => {
    expect(() => new zxc.Seekable(Buffer.alloc(0))).toThrow();
  });

  test("methods after close throw", () => {
    const s = new zxc.Seekable(buildSeekableArchive(buildPayload(4096)));
    s.close();
    expect(() => s.numBlocks()).toThrow();
    // close is idempotent
    expect(() => s.close()).not.toThrow();
  });
});

describe("Seekable: reader callback", () => {
  const payload = buildPayload(128 * 1024);
  const compressed = buildSeekableArchive(payload);

  test("roundtrip via in-memory readAt", () => {
    let calls = 0;
    const reader = {
      size: compressed.length,
      readAt(dst, offset) {
        calls++;
        compressed.copy(dst, 0, offset, offset + dst.length);
      },
    };
    const s = new zxc.Seekable(reader);
    try {
      // 3 reads at open: file header, footer, EOF and SEK headers.
      expect(calls).toBe(3);
      expect(s.decompressedSize()).toBe(payload.length);
      const out = s.decompressRange(0, payload.length);
      expect(Buffer.compare(out, payload)).toBe(0);

      const before = calls;
      const chunk = s.decompressRange(2048, 1024);
      expect(Buffer.compare(chunk, payload.subarray(2048, 2048 + 1024))).toBe(
        0,
      );
      // Single-block sub-range: its seek table entries, then the block.
      expect(calls - before).toBe(2);
    } finally {
      s.close();
    }
  });

  test("readAt throwing maps to decompress_range error", () => {
    let attempted = 0;
    const reader = {
      size: compressed.length,
      readAt(dst, offset) {
        attempted++;
        if (attempted > 3) throw new Error("boom");
        compressed.copy(dst, 0, offset, offset + dst.length);
      },
    };
    const s = new zxc.Seekable(reader);
    try {
      expect(() => s.decompressRange(0, payload.length)).toThrow();
    } finally {
      s.close();
    }
  });

  test("setDict() from readAt during decompressRange is refused", () => {
    let refused = null;
    let s;
    const reader = {
      size: compressed.length,
      readAt(dst, offset) {
        if (refused === null && s) {
          try {
            s.setDict(Buffer.alloc(64, 0x78));
            refused = false;
          } catch {
            refused = true;
          }
        }
        compressed.copy(dst, 0, offset, offset + dst.length);
      },
    };
    s = new zxc.Seekable(reader);
    try {
      const out = s.decompressRange(0, payload.length);
      expect(refused).toBe(true);
      expect(Buffer.compare(out, payload)).toBe(0);
    } finally {
      s.close();
    }
  });

  test("close() from readAt during blockCompressedSize cannot free the handle", () => {
    // blockCompressedSize reads the block's entry through readAt, so it
    // re-enters JS like decompressRange: the same guard must hold, or a
    // close() from the callback frees the handle the library is still using.
    let attempted = 0;
    let s;
    const reader = {
      size: compressed.length,
      readAt(dst, offset) {
        attempted++;
        if (attempted > 3) s.close(); // refused: throws inside the callback
        compressed.copy(dst, 0, offset, offset + dst.length);
      },
    };
    s = new zxc.Seekable(reader);
    try {
      expect(() => s.blockCompressedSize(0)).toThrow();
      // The handle survived the attempt and still answers.
      expect(s.numBlocks()).toBeGreaterThanOrEqual(1);
      expect(s.blockDecompressedSize(0)).toBeGreaterThan(0);
    } finally {
      s.close();
    }
  });

  test("setDict() from readAt during decompressRange is refused", () => {
    // The running decode uses the dictionary a setDict would free.
    let refused = null;
    let s;
    const reader = {
      size: compressed.length,
      readAt(dst, offset) {
        if (refused === null && s) {
          try {
            s.setDict(Buffer.alloc(64, 0x78));
            refused = false;
          } catch {
            refused = true;
          }
        }
        compressed.copy(dst, 0, offset, offset + dst.length);
      },
    };
    s = new zxc.Seekable(reader);
    try {
      const out = s.decompressRange(0, payload.length);
      expect(refused).toBe(true);
      expect(Buffer.compare(out, payload)).toBe(0);
    } finally {
      s.close();
    }
  });

  test("readAt throwing maps to a blockCompressedSize error, not a size of 0", () => {
    let attempted = 0;
    const reader = {
      size: compressed.length,
      readAt(dst, offset) {
        attempted++;
        if (attempted > 3) throw new Error("boom");
        compressed.copy(dst, 0, offset, offset + dst.length);
      },
    };
    const s = new zxc.Seekable(reader);
    try {
      expect(() => s.blockCompressedSize(0)).toThrow();
    } finally {
      s.close();
    }
  });

  test("rejects missing size / readAt", () => {
    expect(() => new zxc.Seekable({})).toThrow();
    expect(() => new zxc.Seekable({ size: 100 })).toThrow();
    expect(() => new zxc.Seekable({ readAt: () => {} })).toThrow();
    expect(() => new zxc.Seekable({ size: 0, readAt: () => {} })).toThrow();
  });

  test("rejects garbage reader", () => {
    const reader = {
      size: 64,
      readAt(dst /* , offset */) {
        dst.fill(0);
      },
    };
    expect(() => new zxc.Seekable(reader)).toThrow();
  });
});

describe("Seekable: low-level seek table helpers", () => {
  test("seekTableSize / writeSeekTable round trip", () => {
    const compSizes = [128, 256, 200, 4];
    const sz = zxc.seekTableSize(compSizes.length);
    expect(sz).toBeGreaterThan(0);
    const buf = zxc.writeSeekTable(compSizes);
    expect(buf.length).toBe(sz);
  });

  test("writeSeekTable rejects empty array", () => {
    expect(() => zxc.writeSeekTable([])).toThrow();
  });
});

describe("Seekable: checksum switch", () => {
  const payload = buildPayload(256 * 1024);

  test("the switch toggles cleanly at any time on an intact archive", () => {
    const s = new zxc.Seekable(buildSeekableArchive(payload));
    try {
      expect(s.decompressRange(0, 512)).toEqual(payload.subarray(0, 512));
      s.setChecksum(false);
      expect(s.decompressRange(0, 512)).toEqual(payload.subarray(0, 512));
      s.setChecksum(true);
      expect(s.decompressRange(0, 512)).toEqual(payload.subarray(0, 512));
    } finally {
      s.close();
    }
  });

  // Incompressible on purpose: a RAW block memcpys a flipped byte straight
  // through, so the decode succeeds and only the checksum objects. A
  // compressible payload would pin this test to today's encoder output.
  test("a corrupted block is silent without verification", () => {
    let rng = 0x2e5b9a17;
    const raw = Buffer.alloc(256 * 1024);
    for (let i = 0; i < raw.length; i++) {
      rng = (Math.imul(rng, 1103515245) + 12345) >>> 0;
      raw[i] = (rng >>> 16) & 0xff;
    }
    const bad = Buffer.from(
      zxc.compress(raw, { seekable: true, checksum: true }),
    );
    expect(bad[16]).toBe(0); // block 0 must be RAW
    bad[16 + 8 + 4] ^= 0xff;

    const on = new zxc.Seekable(bad);
    try {
      on.setChecksum(true);
      expect(() => on.decompressRange(0, 512)).toThrow(/BAD_CHECKSUM/);
    } finally {
      on.close();
    }

    // Default is off: bytes come back, wrong, with no error.
    const off = new zxc.Seekable(bad);
    try {
      expect(off.decompressRange(0, 512)).not.toEqual(raw.subarray(0, 512));
    } finally {
      off.close();
    }
  });

  test("setChecksum rejects a non-boolean", () => {
    const s = new zxc.Seekable(buildSeekableArchive(payload));
    try {
      expect(() => s.setChecksum("yes")).toThrow(TypeError);
    } finally {
      s.close();
    }
  });
});
